runtime/src/accounts_index.rs

```diff
@@ -220,9 +220,124 @@ impl<T: 'static + Clone> AccountsIndex<T> {
+        // `R_new` is an ancestor of `B`, and `R < R_new < B`, then B.ancestors.contains(R_new)`.
+        //
+        // Now until the next `set_root`, any new banks constructed from `new_from_parent` will
+        // also have `max_root == R_new` in their ancestor set, so the claim holds for those descendants
+        // as well. Once the next `set_root` happens, we once again update `max_root` and the same
+        // inductive argument can be applied again to show the claim holds.
+
+        // Check that the `max_root` is present in `ancestors`. From the proof above, if
+        // `max_root` is not present in `ancestors`, this means the bank `B` with the
+        // given `ancestors` is not descended from `max_root, which means
+        // either:
+        // 1) `B` is on a different fork or
+        // 2) `B` is an ancestor of `max_root`.
+        // In both cases we can ignore the given ancestors and instead just rely on the roots
+        // present as `max_root` indicates the roots present in the index are more up to date
+        // than the ancestors given.
+        let empty = HashMap::new();
+        let ancestors = if ancestors.contains_key(&max_root) {
+            ancestors
+        } else {
+            /*
+            This takes of edge cases like:
+
+            Diagram 1:
+
+                        slot 0
+                          |
+                        slot 1
+                      /        \
+                 slot 2         |
+                    |       slot 3 (max root)
+            slot 4 (scan)
+
+            By the time the scan on slot 4 is called, slot 2 may already have been
+            cleaned by a clean on slot 3, but slot 4 may not have been cleaned.
+            The state in slot 2 would have been purged and is not saved in any roots.
+            In this case, a scan on slot 4 wouldn't accurately reflect the state when bank 4
+            was frozen. In cases like this, we default to a scan on the latest roots by
+            removing all `ancestors`.
+            */
+            &empty
+        };
+
+        /*
+        Now there are two cases, either `ancestors` is empty or nonempty:
+
+        1) If ancestors is empty, then this is the same as a scan on a rooted bank,
+        and `ongoing_scan_roots` provides protection against cleanup of roots necessary
+        for the scan, and  passing `Some(max_root)` to `do_scan_accounts()` ensures newer
+        roots don't appear in the scan.
+
+        2) If ancestors is non-empty, then from the `ancestors_contains(&max_root)` above, we know
+        that the fork structure must look something like:
+
+        Diagram 2:
+
+                Build fork structure:
+                        slot 0
+                          |
+                        slot 1
+                    /            \
+        slot 2 (max_root)         |
+                |            slot 3 (potential newer max root)
```

Comment:
slot 2 is rooted, right? how could slot 3 be rooted which doesn't chain to slot 2?

Reply:
yeah you're right, updated and changed the `max_root` to be at slot 1, thanks!